### PR TITLE
statistics: fix auto analyze for global index

### LINKED ISSUE #47539: autoanalyze failed when a table has global index, analyze version = 1
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```SQL
set global tidb_analyze_version = 1;
use test;
CREATE TABLE t ( a int, b int, c int default 0)
  PARTITION BY RANGE (a) (
  PARTITION p0 VALUES LESS THAN (10),
  PARTITION p1 VALUES LESS THAN (20),
  PARTITION p2 VALUES LESS THAN (30),
  PARTITION p3 VALUES LESS THAN (40));
INSERT INTO t(a, b) values(1, 1), (2, 2), (3, 3), (15, 15), (25, 25), (35, 35);
ALTER TABLE t ADD UNIQUE INDEX idx(b);
set global tidb_partition_prune_mode = "dynamic";
```

### 2. What did you expect to see? (Required)
auto analyze run successfully.

### 3. What did you see instead (Required)
auto analyze run failed, global index should be handle specifically.
image

### 4. What is your TiDB version? (Required)
image

### PR DESCRIPTION
**Pull request by L-maple**
### What problem does this PR solve?

Issue Number: close #47539

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by qw4990** [comment]
@hawkingrei PTAL

File: pkg/planner/core/planbuilder.go [hunk 1]
```diff
@@ -2952,8 +2952,11 @@ func (b *PlanBuilder) buildAnalyzeAllIndex(as *ast.AnalyzeTableStmt, opts map[as
 	return p, nil
 }
 
-func generateIndexTasks(idx *model.IndexInfo, as *ast.AnalyzeTableStmt, tblInfo *model.TableInfo, names []string, physicalIDs []int64, version int) []AnalyzeIndexTask {
-	if idx.Global {
+func generateIndexTasks(idx *model.IndexInfo, as *ast.AnalyzeTableStmt, tbl *ast.TableName, names []string, physicalIDs []int64, version int) []AnalyzeIndexTask {
+	tblInfo := tbl.TableInfo
+
+	isAnalyzeTable := len(as.PartitionNames) == 0
```
Comment (hawkingrei) [inline pkg/planner/core/planbuilder.go:2958]: Can we called ```isPartitionTable```?

**Comment by hawkingrei** [comment]
@mapleFU Please run ```make bazel_prepare``` to update bazel config.

**Review by hawkingrei — APPROVED** [review]

**Comment by winoros** [comment]
@L-maple, I think the changes in this pull will not make the global index's statistics collecting a separate process from the merging global stats?

**Comment by L-maple** [comment]
> @L-maple, I think the changes in this pull will make the global index's statistics collecting a separate process from the merging global stats?

@winoros Thanks, I revise the PR again for merging global stats compatibility.
in `handleGlobalStats`, `if results.TableID.IsPartitionTable() && needGlobalStats {` will check the global index is not needed to be merged.

**Comment by L-maple** [comment]
@winoros Hi, do you have any other advice？This PR can be reiviewed again.

**Comment by winoros** [comment]
> @winoros Hi, do you have any other advice？This PR can be reiviewed again.

It needs some time to discuss the behavior. How could I reach you via some IM?

**Comment by L-maple** [comment]
Thanks for your feedback，my wechat ID is：lcj960929（刘常杰）
Maybe I should add some comments for global index statistics in the code.



---- Replied Message ----
| From | Yiding ***@***.***> |
| Date | 11/13/2023 22:47 |
| To | ***@***.***> |
| Cc | Changjie ***@***.***>***@***.***> |
| Subject | Re: [pingcap/tidb] statistics: fix auto analyze for global index (PR #47754) |

@winoros Hi, do you have any other advice？This PR can be reiviewed again.

It needs some time to discuss the behavior. How could I reach you via some IM?

**Pull request closed without merging**